src/registry/artifactory.rs

```diff
@@ -26,191 +27,166 @@ pub struct Artifactory {
-        );
-
-        let version = dependency
-            .manifest
-            .version
-            .comparators
-            .first()
-            .wrap_err("internal error")?;
-
-        ensure!(
-            version.op == semver::Op::Exact && version.minor.is_some() && version.patch.is_some(),
-            "artifactory only support pinned dependencies"
-        );
+    /// Publishes a package to artifactory
+    pub async fn publish(&self, package: Package, repository: String) -> miette::Result<()> {
+        let artifact_uri: Url = format!(
+            "{}/{}/{}/{}-{}.tgz",
+            self.registry,
+            repository,
+            package.name(),
+            package.name(),
+            package.version(),
+        )
+        .parse()
+        .into_diagnostic()
+        .wrap_err(miette!(
+            "unexpected error: failed to construct artifact URL"
+        ))?;
+
+        let _ = self
+            .new_request(Method::PUT, artifact_uri)
+            .body(package.tgz.clone())
+            .send()
+            .await?;
 
-        let version = format!(
-            "{}.{}.{}{}",
-            version.major,
-            version.minor.wrap_err("internal error")?,
-            version.patch.wrap_err("internal error")?,
-            if version.pre.is_empty() {
-                "".to_owned()
-            } else {
-                format!("-{}", version.pre)
-            }
+        tracing::info!(
+            ":: published {}/{}@{}",
+            repository,
+            package.name(),
+            package.version()
         );
 
-        let artifact_url: Url = {
-            let path = dependency.manifest.registry.path().to_owned();
+        Ok(())
+    }
+}
 
-            let mut url = dependency.manifest.registry.clone();
-            url.set_path(&format!(
-                "{}/{}/{}/{}-{}.tgz",
-                path,
-                dependency.manifest.repository,
-                dependency.package,
-                dependency.package,
-                version
-            ));
+struct RequestBuilder(reqwest::RequestBuilder);
 
-            url.into()
-        };
+impl RequestBuilder {
+    fn new(client: reqwest::Client, method: reqwest::Method, url: Url) -> Self {
+        Self(client.request(method, url))
+    }
 
-        tracing::debug!("Hitting download URL: {artifact_url}");
+    fn auth(mut self, token: String) -> Self {
+        self.0 = self.0.bearer_auth(token);
+        self
+    }
 
-        let mut request = self.client.get(artifact_url);
+    fn body(mut self, payload: impl Into<Body>) -> Self {
+        self.0 = self.0.body(payload);
+        self
+    }
 
-        if let Some(token) = &self.token {
-            request = request.bearer_auth(token);
-        }
+    async fn send(self) -> miette::Result<ValidatedResponse> {
+        self.0.send().await.into_diagnostic()?.try_into()
+    }
+}
 
-        let response = request.send().await?;
+struct ValidatedResponse(reqwest::Response);
```

Comment:
i would impl also `into_inner`, its common for wrappers

Reply:
I impl `into()` instead, isn't that sufficient?

Reply:
I can replace the `From<ValidatedResponse> for reqwest::Response` implementation if you want, but it's pretty much either/or, otherwise we end up with dead code.

Reply:
#103